src/csharp/Microsoft.Spark/Pyrolite/BufferUnpickler.cs

```diff
@@ -0,0 +1,859 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// See the LICENSE file in the project root for more information.
+
+/* part of Pyrolite, by Irmen de Jong (irmen@razorvine.net) */
+
+using System;
+using System.Buffers.Binary;
+using System.Buffers.Text;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Text;
+using Razorvine.Pickle.Objects;
+
+namespace Razorvine.Pickle
+{
+    /// <summary>
+    /// Unpickles an object graph from a pickle data memory buffer. Supports all pickle protocol versions.
+    /// Maps the python objects on the corresponding java equivalents or similar types.
+    /// This class is NOT threadsafe! (Don't use the same unpickler from different threads)
+    /// See the README.txt for a table with the type mappings.
+    /// </summary>
+    [SuppressMessage("ReSharper", "InconsistentNaming")]
+    [SuppressMessage("ReSharper", "MemberCanBePrivate.Global")]
+    [SuppressMessage("ReSharper", "InvertIf")]
+    internal class BufferUnpickler
```

Comment:
Do we really need to duplicate the whole class? There appears to be a lot of duplication between this and Unpickler.cs.

One pattern I used in the Arrow repo was to have the "public" class be the same, and then internally under the covers used a "strategy" pattern for the "implementation" over a Stream or Memory.

Public class with 2 constructors:
[https://github.com/apache/arrow/blob/master/csharp/src/Apache.Arrow/Ipc/ArrowStreamReader.cs#L37-L48](https://github.com/apache/arrow/blob/master/csharp/src/Apache.Arrow/Ipc/ArrowStreamReader.cs#L37-L48)

2 internal implementations:
[https://github.com/apache/arrow/blob/master/csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs](https://github.com/apache/arrow/blob/master/csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs)
[https://github.com/apache/arrow/blob/master/csharp/src/Apache.Arrow/Ipc/ArrowMemoryReaderImplementation.cs](https://github.com/apache/arrow/blob/master/csharp/src/Apache.Arrow/Ipc/ArrowMemoryReaderImplementation.cs)

And an internal base class for shared logic:
[https://github.com/apache/arrow/blob/master/csharp/src/Apache.Arrow/Ipc/ArrowReaderImplementation.cs](https://github.com/apache/arrow/blob/master/csharp/src/Apache.Arrow/Ipc/ArrowReaderImplementation.cs)

thoughts?